### PR TITLE
Components: ComboboxControl supports disabled items 

### LINKED ISSUE #60969: Components: ComboboxControl Item should support `disabled` option
Considering the `disabled` option is supported natively by the `<select>` HTML component for their children, it should similarly be supported by the `<ComboboxControl />` Item component.

[Here is](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/select#advanced_select_with_multiple_features) an example of the `disabled` option for the regular `<select>` HTML element.

### Item API

Once this implementation:

* it would be possible to disable the item by passing the `disabled` property:
* Item is not disabled as default

```jsx
import { ComboboxControl } from '@wordpress/components';
import { useState } from '@wordpress/element';

const options = [
	{
		value: 'small',
		label: 'Small',
	},
	{
		value: 'normal',
		label: 'Normal',
		disabled: true,
	},
	{
		value: 'large',
		label: 'Large',
		disabled: false,
	},
];

function MyComboboxControl() {
	const [ fontSize, setFontSize ] = useState();
	const [ filteredOptions, setFilteredOptions ] = useState( options );
	return (
		<ComboboxControl
			label="Font Size"
			value={ fontSize }
			onChange={ setFontSize }
			options={ filteredOptions }
			onFilterValueChange={ ( inputValue ) =>
				setFilteredOptions(
					options.filter( ( option ) =>
						option.label
							.toLowerCase()
							.startsWith( inputValue.toLowerCase() )
					)
				)
			}
		/>
	);
}
```

### PR DESCRIPTION
**Pull request by retrofox**
## What?

This PR supports disabled items for the `<ComboboxControl / >` component.

## Why?

Because the `disable` option for a Combobox component is a common requirement when using it for the UI.

## How?

The `<ComboboxControl />` now accepts the `disabled` property, which is propagated to the `SuggestionsList` one.
The implementation checks if the option is disabled. If so, it does not allow selection in the respective event handler functions.

Let's remark that the disable options are not selectable (by clicking on it or pressing the `enter` key), but they are discoverable and can make the implementation a11y-friendly.

## Testing Instructions

This PR adds a new story to test the `disable` items:

1. Run the storybook 

```cli
npm run storybook:dev
```

2. Select the ComboboxControl story

```
http://localhost:50240/?path=/story/components-comboboxcontrol--with-disabled-options
```

4. Confirm the disabled items looks like disabled :smart:
5. Confirm it is not possible to select the items by clicking on them
6. Confirm it is possible to "discover" the item by pressing the UP/DOWN arrow
7. Confirm it is not possible to select the items by pressing the `ENTER` key

https://github.com/WordPress/gutenberg/assets/77539/57d42a80-3350-4a18-bc8e-be0508d84600

### CODE REVIEW

**Review by alexstine — COMMENTED** [review]
@retrofox Thanks for the PR. One accessibility note, you should not be restricting arrow navigation for disabled options. By doing that, you are taking the ability away from keyboard-only users to hear if an option is disabled or not. I see you are already using the `aria-disabled` attribute on the disabled options, this is good, but users would likely never hear this announced since you are skipping any index that is disabled on arrow key down.

**Comment by retrofox** [comment]
Hi, @alexstine. Thanks for your feedback.

> ...you should not be restricting arrow navigation for disabled options. ...

Good point. This is something that I've been thinking about, too. However, I followed the behavior of the native element.

https://developer.mozilla.org/en-US/docs/Web/HTML/Element/select#advanced_select_with_multiple_features

It isn't possible to select the disabled option by using the keyboard.

Anyway, we can make them accessible if we consider this to be the proper behavior.

**Comment by alexstine** [comment]
@retrofox I think at minimum, the option should at least be discoverable even if it is announced as disabled.

Thoughts @afercia @joedolson?

So many bad examples around the web. I fail to understand how people see the experience as fair or equitable when we make things visible in the UI visually but not readable allowed for VI/blind users.

**Comment by retrofox** [comment]
> I think at minimum, the option should at least be discoverable even if it is announced as disabled.

💯 sgtm

**Comment by afercia** [comment]
> I think at minimum, the option should at least be discoverable even if it is announced as disabled

I agree the 'not seleftable' options need to be discoverable. Any UI must provide information parity for all users.
- Sighted users can actually _see_ the non-selectable option and understand it's 'disabled' as in: not selectable.
- Non sighted users must be provided with an equivalent information. As such, they need to be able to navigate to the non-selectable option.

File: packages/components/src/form-token-field/suggestions-list.tsx
Comment (mirka) [inline packages/components/src/form-token-field/suggestions-list.tsx]: Is the `.is-disabled` class necessary? Couldn't we just use `[aria-disabled="true"]` as the selector?

File: packages/components/src/form-token-field/suggestions-list.tsx
Comment (mirka) [inline packages/components/src/form-token-field/suggestions-list.tsx]: Haven't tested, but would it be easier to set `pointer-events: none` in CSS so disabled options don't trigger click events in the first place? That might be more robust against future code changes.

File: packages/components/src/form-token-field/style.scss
Comment (mirka) [inline packages/components/src/form-token-field/style.scss]: I think we want an explicit `color` here (`$gray-600`, I think) rather than an opacity, to ensure that the text meets the contrast threshold.

**Review by mirka — COMMENTED** [review]
@WordPress/gutenberg-design Could we have a quick design check on the styling for the disabled options?

File: packages/components/src/form-token-field/suggestions-list.tsx
Comment (retrofox) [reply packages/components/src/form-token-field/suggestions-list.tsx]: yes, we can. Let's remove it.

File: packages/components/src/form-token-field/suggestions-list.tsx
Comment (retrofox) [reply packages/components/src/form-token-field/suggestions-list.tsx]: nice one. Let me try it

File: packages/components/src/form-token-field/style.scss
Comment (retrofox) [reply packages/components/src/form-token-field/style.scss]: I followed other implementations. Let's try it

File: packages/components/src/form-token-field/suggestions-list.tsx
Comment (retrofox) [reply packages/components/src/form-token-field/suggestions-list.tsx]: https://github.com/WordPress/gutenberg/pull/61294/commits/b03f3f90b85c2f65e35dfdcfa03edb54c2b3198f

File: packages/components/src/form-token-field/style.scss
Comment (retrofox) [reply packages/components/src/form-token-field/style.scss]: https://github.com/WordPress/gutenberg/pull/61294/commits/87290d852245cf3fa1390d060841744648c875d5

**Comment by jasmussen** [comment]
> Could we have a quick design check on the styling for the disabled options?

I'd want to defer to best practices here. At a glance the low-opacity/grayscale makes it look disabled, but I'm unfamiliar with practices for selected-disabled-item styles. If what is present in this branch passes instincts, let's try that.

**Comment by retrofox** [comment]
> > Could we have a quick design check on the styling for the disabled options?
> 
> I'd want to defer to best practices here. At a glance the low-opacity/grayscale makes it look disabled, but I'm unfamiliar with practices for selected-disabled-item styles. If what is present in this branch passes instincts, let's try that.

The following shows the current style based on `color` and `background-color` values:

`Algeria` is disabled:

image

`Algeria` is disabled and selected:

image

I decided to apply the `$grey-100` background color in the same spirit as `aria-disable`, to let the user know that the current (disabled) option is focused. However, it isn't possible to select it.

**Comment by jasmussen** [comment]
Again, will defer to others. To me this is not a design issue, this is a question of: what are the constraints we are reaching for as to how this component should behave in an accessible way. If the gray works, go for it. An alternative is to keep the blue selection style, but instead show a semi-transparent white text-color for the disabled item when it's selected — my instincts would be that this would be less jumpy as you navigate.

**Comment by retrofox** [comment]
Thanks, Joen

>  An alternative is to keep the blue selection style, but instead show a semi-transparent white text-color for the disabled item when it's selected — my instincts would be that this would be less jumpy as you navigate.

I do agree.

**Comment by retrofox** [comment]
> An alternative is to keep the blue selection style, but instead show a semi-transparent white text-color for the disabled item when it's selected


https://github.com/WordPress/gutenberg/assets/77539/ec4ec756-62dd-469f-849a-e78271a845f3

File: packages/components/src/form-token-field/style.scss
Comment (retrofox) [inline packages/components/src/form-token-field/style.scss]: this line triggers lint-style error:

```
✖ wp-scripts lint-style found some errors. Please fix them and try committing again.
 
packages/components/src/form-token-field/style.scss
191:16  ✖  Unexpected whitespace after "("   function-parentheses-space-inside
191:35  ✖  Unexpected whitespace before ")"  function-parentheses-space-inside
```

I committed this change with force

File: packages/components/src/combobox-control/stories/index.story.tsx
Comment (tyxla) [inline packages/components/src/combobox-control/stories/index.story.tsx]: Could this also use a comment (that we're intentionally making the first and every 3rd item disabled)?

File: packages/components/src/form-token-field/style.scss
Comment (tyxla) [reply packages/components/src/form-token-field/style.scss]: Yes, the usual extra spacing based on WP coding style doesn't really apply to CSS functions. Perhaps we can just satisfy the linter?

```suggestion
			color: rgba(255, 255, 255, 0.5);
```

File: packages/components/src/form-token-field/suggestions-list.tsx
Comment (tyxla) [inline packages/components/src/form-token-field/suggestions-list.tsx:110]: Do we need the `object` type check? Calling `stringVal?.disabled` should be just fine for a string, it will just return `undefined`.

File: packages/components/src/combobox-control/index.tsx
Comment (tyxla) [inline packages/components/src/combobox-control/index.tsx:81]: As a new API it definitely makes sense to include a CHANGELOG entry for this PR.

**Review by tyxla — COMMENTED** [review]
This looks good to me, code-wise. 

I've left a few optional suggestions, among which there are a couple that should also satisfy all checks. 

I'll let design folks do the final approval since there was some design feedback.

File: packages/components/src/combobox-control/index.tsx
Comment (retrofox) [reply packages/components/src/combobox-control/index.tsx:81]: I tried it [here](https://github.com/WordPress/gutenberg/pull/61294/commits/0e61a05621444f1dc57b83e13ac604f89235425d). Am I missing something?

File: packages/components/src/form-token-field/suggestions-list.tsx
Comment (retrofox) [reply packages/components/src/form-token-field/suggestions-list.tsx:110]: `stringVal?.disabled`, considering `stringVal` is a String, generates a TS error:

```
Property 'disabled' does not exist on type 'string | { value: string; disabled?: boolean | undefined; }'.
  Property 'disabled' does not exist on type 'string'.ts.
```

So we need to check the type before trying to access to the `disabled` property

File: packages/components/src/combobox-control/index.tsx
Comment (tyxla) [reply packages/components/src/combobox-control/index.tsx:81]: That's correct. It just needs to be moved to "Unreleased" though, that's why the check is failing.

**Comment by jameskoster** [comment]
In a lot of Combobox examples I see, disabled options do tend to lighten the background as well as the text. I presume this is to create adequate visual contrast between disabled/enabled options. Perhaps it would work a bit better if the background was a lighter shade of the admin theme color rather than gray?

disabled

**Comment by retrofox** [comment]
Hi, James. Thanks for your input.

> In a lot of Combobox examples I see, disabled options do tend to lighten the background as well as the text. 

Could you share a few UI design Combobox examples where disabled options lighten the background and text? I did the same, and considering there are many options... it doesn't help to define which approach we should take :-(
However, we should probably follow it, considering (I think) that we'd like to align more with Ariakit. 
Ariakit applies simply `opacity: 0.25;`:

Screenshot 2024-05-08 at 11 23 08


[example](https://ariakit.org/examples/menu-nested)

**Comment by jameskoster** [comment]
That example is a `menu` rather than a `combobox`, so perhaps _slightly_ different? Curiously comboboxes in ariakit don't seem to have any visual treatment for disabled options, which is a bit puzzling.

The example I looked at earlier was [this one](https://citizensadvice.github.io/react-combo-boxes/combo_box/disabled_option/index.html). There are more [here](https://designsystems.surf/components/combobox). 

Generally disabled options don't share a background treatment with enabled ones. Oftentimes disabled options have no hover effect at all, which could also work here imo.

**Comment by retrofox** [comment]
> The example I looked at earlier was [this one](https://citizensadvice.github.io/react-combo-boxes/combo_box/disabled_option/index.html). There are more [here](https://designsystems.surf/components/combobox).

👍 

> Generally disabled options don't share a background treatment with enabled ones. Oftentimes disabled options have no hover effect at all, which could also work here imo.

having no effect, I think, makes sense for `disabled` options. For `aria-disabled` maybe it doesn't?
Anyway, I'm open to any implement we want to do

**Comment by jameskoster** [comment]
> having no effect, I think, makes sense for disabled options. For aria-disabled maybe it doesn't?

Yes that's a good point. For sighted keyboard users it does make sense for there to be a background change when an aria-disabled option is highlighted 👍 

One small tweak to the mockup above, we might apply `cursor: default` to the disabled option, to further illustrate the option is not selectable to mouse users. This matches disabled buttons.

Disabled option in a combobox' listbox

**Comment by retrofox** [comment]
> One small tweak to the mockup above, we might apply `cursor: default` to the disabled option, to further illustrate the option is not selectable to mouse users. This matches disabled buttons.

The mouse pointer is not captured by my screenshots 😅 , but it looks similar to what you shared (believe me :-D ) , with an exception: The style is not applied to the disabled option when the mouse hovers. This is because it applies `pointer-events: none;`

**Comment by jameskoster** [comment]
> This is because it applies pointer-events: none;

Oh is that necessary? I think I'd expect the hover and highlight (via keyboard) visuals to match.

I'm still seeing the solid blue background on disabled options:

Screenshot 2024-05-09 at 09 33 54

For the disabled option background we could potentially use `rgba(var(--wp-admin-theme-color--rgb), 0.04)` which matches tertiary buttons.

**Comment by retrofox** [comment]
> Oh is that necessary? I think I'd expect the hover and highlight (via keyboard) visuals to match.

This approach, in my view, is not just good but possibly the best way to handle the disabled state. It's a change proposed by our development team and is also a prevalent practice in other libraries like Ariaakit.


> I'm still seeing the solid blue background on disabled options:
> For the disabled option background we could potentially use rgba(var(--wp-admin-theme-color--rgb), 0.04) which matches tertiary buttons.

👍

**Comment by retrofox** [comment]
item selected | disabled item focused
-------|--------
image | image

**Review by jameskoster — APPROVED** [review]
Seems in a good spot to me :)

**Comment by retrofox** [comment]
> Seems in a good spot to me :)

Thanks for your help. Super valuable. Let's wait for the other 👍s!

**Comment by retrofox** [comment]
> Seems in a good spot to me :)

Same :-D Merging...

File: packages/components/src/utils/theme-variables.scss
Comment (ciampo) [inline packages/components/src/utils/theme-variables.scss:10]: While reviewing `--wp-components-color-*` variables, @mirka and I came across this change that was added just before merging the PR and escaped our review.

- the `--wp-components-color-accent--rgb` custom property does not exist (we don't have any `--rgb` variants);
- the `$components-color-accent-transparent-40` SASS variable doesn't need to exist as a common variable, as it can be derived directly where it's needed via the `color-mix()` function
- it doesn't work on dark backgrounds (ie dark mode)

We should remove this variable and figure out an alternative solution that only makes use of existing variables and works with all possible theme combinations.

**Pull request merged**
